NBitcoin.Tests/NetworkAddressTests.cs

```diff
@@ -0,0 +1,567 @@
+
+			// TORv3
+			var torv3_addr = "pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion";
+			Assert.True(addr.SetSpecial(torv3_addr));
+			Assert.True(addr.IsTor);
+
+			Assert.False(addr.IsAddrV1Compatible);
+			Assert.Equal(addr.ToAddressString(), torv3_addr);
+
+			// TORv3, broken, with wrong checksum
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscsad.onion"));
+
+			// TORv3, broken, with wrong version
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscrye.onion"));
+
+			// TORv3, malicious
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd\0wtf.onion"));
+
+			// TOR, bogus length
+			Assert.False(addr.SetSpecial("mfrggzak.onion"));
+
+			// TOR, invalid base32
+			Assert.False(addr.SetSpecial("mf*g zak.onion"));
+
+			// I2P
+			var i2p_addr = "udhdrtrcetjm5sxzskjyr5ztpeszydbh4dpl3pl4utgqqw2v4jna.b32.i2p";
+			Assert.True(addr.SetSpecial(i2p_addr));
+			Assert.True(addr.IsI2P);
+			Assert.False(addr.IsAddrV1Compatible);
+			Assert.Equal(addr.ToAddressString(), i2p_addr);
+
+			// I2P, malicious
+			Assert.False(addr.SetSpecial("udhdrtrcetjm5sxzskjyr5ztpeszydbh4dpl3pl4utgqqw2v4jna\0wtf.b32.i2p"));
+
+			// I2P, valid but unsupported 
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscsad.b32.i2p"));
+
+			// I2P, invalid base32
+			Assert.False(addr.SetSpecial("tp*szydbh4dp.b32.i2p"));
+
+			// General validations
+			Assert.False(addr.SetSpecial(".onion"));
+			Assert.False(addr.SetSpecial(""));
+		}
+
+		[Fact]
+		[Trait("UnitTest", "UnitTest")]
+		public void SerializeV1()
+		{
+			var addr = new NetworkAddress();
+			var mem = new MemoryStream();
+			var stream = new BitcoinStream(mem, true);
+
+			string AddressHex()
+			{
+				var arr = mem.ToArray();
+				return Encoders.Hex.EncodeData(arr, 8, 16);
+			}
+
+			// serialize method produces an address in v2 format.
+			stream.Type = SerializationType.Hash;
+
+			stream.ReadWrite(addr);
+			Assert.Equal("00000000000000000000000000000000", AddressHex());
+
+			mem.Clear();
+			addr = new NetworkAddress(IPAddress.Parse("1.2.3.4"));
+			stream.ReadWrite(addr);
+			Assert.Equal("00000000000000000000ffff01020304", AddressHex());
+
+			mem.Clear();
+			addr = new NetworkAddress(IPAddress.Parse("1a1b:2a2b:3a3b:4a4b:5a5b:6a6b:7a7b:8a8b"));
+			stream.ReadWrite(addr);
+			Assert.Equal("1a1b2a2b3a3b4a4b5a5b6a6b7a7b8a8b", AddressHex());
+
+			mem.Clear();
+			addr = new NetworkAddress();
+			addr.SetSpecial("6hzph5hv6337r6p2.onion");
+			stream.ReadWrite(addr);
+			Assert.Equal("fd87d87eeb43f1f2f3f4f5f6f7f8f9fa", AddressHex());
+
+			mem.Clear();
+			addr = new NetworkAddress();
+			addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion");
+			stream.ReadWrite(addr);
+			Assert.Equal("00000000000000000000000000000000", AddressHex());
+		}
+
+		[Fact]
```

Comment:
Helper function `MakeNewStream` can be used here:

```csharp
[Fact]
[Trait("UnitTest", "UnitTest")]
public void UnserializeV2()
{
	BitcoinStream MakeNewStream(byte[] payload)
	{
		return new BitcoinStream(payload)
		{
			ProtocolVersion = NetworkAddress.AddrV2Format,
			Type = SerializationType.Hash
		};
	}

	var addr = new NetworkAddress();

	// Valid IPv4.
	var payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"01" +         // network type (IPv4)
		"04" +         // address length
		"01020304" +   // address
		"0000");       // port
	var stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsIPv4);
	Assert.True(addr.IsAddrV1Compatible);
	Assert.Equal("1.2.3.4", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv4, valid length but address itself is shorter.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"01" +         // network type (IPv4)
		"04" +         // address length
		"0102" +       // address
		"0000");       // port
	stream = MakeNewStream(payload);
	Assert.Throws<EndOfStreamException>(() => stream.ReadWrite(ref addr));
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv4, with bogus length.
	// Invalid IPv4, valid length but address itself is shorter.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"01" +         // network type (IPv4)
		"05" +         // address length
		"0102030400"); // address
	stream = MakeNewStream(payload);
	var ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 IPv4 address is 5 bytes long. Expected length is 4 bytes.", ex.Message);

	// Valid IPv6.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"10" +         // address length
		"0102030405060708090a0b0c0d0e0f10" +  // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);

	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsIPv6);
	Assert.True(addr.IsAddrV1Compatible);
	Assert.Equal("102:304:506:708:90a:b0c:d0e:f10", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv6, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"04" +         // address length
		"00000000");   // address
	stream = MakeNewStream(payload);

	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 IPv6 address is 4 bytes long. Expected length is 16 bytes.", ex.Message);

	// Invalid IPv6, contains embedded IPv4.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"10" +         // address length
		"00000000000000000000ffff01020304" +  // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);

	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv6, contains embedded TORv2.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"10" +         // address length
		"fd87d87eeb430102030405060708090a" +   // address
		"0000");       // port

	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Valid TORv2.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"03" +         // network type (TORv2)
		"0a" +         // address length
		"f1f2f3f4f5f6f7f8f9fa" +  // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsTor);
	Assert.True(addr.IsAddrV1Compatible);
	Assert.Equal("6hzph5hv6337r6p2.onion", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid TORv2, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"03" +         // network type (TORv2)
		"07" +         // address length
		"00000000000000");   // address

	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 TORv2 address is 7 bytes long. Expected length is 10 bytes.", ex.Message);

	// Valid TORv3.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"04" +         // network type (TORv3)
		"20" +         // address length
		"79bcc625184b05194975c28b66b66b0469f7f6556fb1ac3189a79b40dda32f1f" +   // address
		"0000");       // port

	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsTor);
	Assert.False(addr.IsAddrV1Compatible);
	Assert.Equal("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid TORv3, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"04" +         // network type (TORv3)
		"00" +         // address length
		"00");        // address

	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 TORv3 address is 0 bytes long. Expected length is 32 bytes.", ex.Message);
	Assert.True(stream.Inner.Length != stream.Inner.Position);

	// Valid I2P.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"05" +         // network type (I2P)
		"20" +         // address length
		"a2894dabaec08c0051a481a6dac88b64f98232ae42d4b6fd2fa81952dfe36a87" + // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsI2P);
	Assert.False(addr.IsAddrV1Compatible);
	Assert.Equal("ukeu3k5oycgaauneqgtnvselmt4yemvoilkln7jpvamvfx7dnkdq.b32.i2p", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid I2P, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"05" +         // network type (I2P)
		"03" +         // address length
		"000000");     // address

	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 I2P address is 3 bytes long. Expected length is 32 bytes.", ex.Message);

	// Valid CJDNS.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"06" +         // network type (CJDNS)
		"10" +         // address length
		"fc000001000200030004000500060007" + // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsCjdns);
	Assert.False(addr.IsAddrV1Compatible);
	Assert.Equal("fc00:1:2:3:4:5:6:7", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid CJDNS, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"06" +         // network type (CJDNS)
		"01" +         // address length
		"00");         // address
	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 Cjdns address is 1 bytes long. Expected length is 16 bytes.", ex.Message);

	// Unknown, with extreme length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"aa" +         // network type (unknown)
		"fe00000002" + // address length (CompactSize's MAX_SIZE)
		"01020304050607"); // address
	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentOutOfRangeException>(() => stream.ReadWrite(ref addr));
	Assert.Contains("Array size too big", ex.Message);
	Assert.True(stream.Inner.Length != stream.Inner.Position);

	// Unknown, with reasonable length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"aa" +         // network type (unknown)
		"04" +         // address length
		"01020304" +   // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Unknown, with zero length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"aa" +         // network type (unknown)
		"00" +         // address length
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);
}
```

Reply:
I like it.